lib/datadog/core/configuration/components.rb

```diff
@@ -57,17 +57,35 @@ def build_runtime_metrics_worker(settings)
 
           def build_telemetry(settings, agent_settings, logger)
             enabled = settings.telemetry.enabled
-            if agent_settings.adapter != Datadog::Core::Configuration::Ext::Agent::HTTP::ADAPTER
+            agentless_enabled = settings.telemetry.agentless_enabled
+
+            if !agentless_enabled && agent_settings.adapter != Datadog::Core::Configuration::Ext::Agent::HTTP::ADAPTER
               enabled = false
               logger.debug { "Telemetry disabled. Agent network adapter not supported: #{agent_settings.adapter}" }
             end
 
+            if agentless_enabled && settings.api_key.nil?
+              enabled = false
+              logger.debug { 'Telemetry disabled. Agentless telemetry requires an DD_API_KEY variable to be set.' }
+            end
+
+            transport = if agentless_enabled
+                          Datadog::Core::Telemetry::Http::Transport.build_agentless_transport(
+                            api_key: settings.api_key,
+                            dd_site: settings.site
+                          )
+                        else
+                          Datadog::Core::Telemetry::Http::Transport.build_agent_transport
+                        end
```

Comment:
Minor: There's a lot of telemetry-specific logic in here. I wonder if it would make sense to embed it directly on the telemetry transport or components. E.g. It's a bit "surprising" how much the core configuration components needs to know about these things right now.

(But I do consider it minor, if you don't really want to touch it :D )

Reply:
moved to a factory method in Telemetry::Component